packages/paymasters/contracts/PermitERC20UniswapV3Paymaster.sol

```diff
@@ -147,33 +208,110 @@ contract PermitERC20UniswapV3Paymaster is BasePaymaster, ERC2771Recipient {
+        if (hubBalance >= minHubBalance + ethActualCharge) {
+            return;
+        }
+        uint256 depositAmount = targetHubBalance - hubBalance + ethActualCharge;
+        _refillHubDeposit(depositAmount);
+    }
+
+    function _refillHubDeposit(uint256 depositAmount) private {
+        uint256 balance = address(this).balance;
+        uint256 amountSwapped = 0;
+        if (balance < depositAmount) {
+            for (uint256 i = 0; i < tokens.length && balance + amountSwapped < depositAmount; i++) {
+                IERC20Metadata tokenIn = tokens[i];
+                uint256 tokenBalance = tokenIn.balanceOf(address(this));
+                if (tokenBalance > 0) {
+                    uint256 quote = uint256(priceFeeds[tokenIn].latestAnswer());
+                    uint256 divisor = priceDivisors[tokenIn];
+                    uint24 poolFee = uniswapPoolFees[tokenIn];
+                    uint256 amountOutMin = _tokenToWei(tokenBalance, divisor, quote) * 99 / 100;
+                    uint256 amountOut = UniswapV3Helper.swapToToken(
+                        address(tokenIn),
+                        address(weth),
+                        tokenBalance,
+                        amountOutMin,
+                        poolFee,
+                        uniswap
+                    );
+                    amountSwapped += amountOut;
+                }
+            }
+            UniswapV3Helper.unwrapWeth(uniswap, amountSwapped);
+        }
+        if (balance + amountSwapped < depositAmount) {
+            emit FundingNeeded();
+            depositAmount = balance + amountSwapped;
+        }
+        relayHub.depositFor{value : depositAmount}(address(this));
+    }
+
+    function _withdrawToOwnerIfNeeded() private {
+        uint256 hubBalance = relayHub.balanceOf(address(this));
+        if (hubBalance >= minWithdrawalAmount + targetHubBalance) {
+        relayHub.withdraw(payable(owner()), hubBalance - targetHubBalance);
+        }
+    }
+
+    function _getTokenFromPaymasterData(bytes calldata paymasterData) internal pure returns (IERC20Metadata) {
+        return IERC20Metadata(address(bytes20(paymasterData[paymasterData.length - 20:])));
+    }
+
+    function addPaymasterFee(uint256 charge) public view returns (uint256) {
+        return charge * (100 + paymasterFee) / 100;
     }
 
     // as this Paymaster already has a permission from a user to operate the tokens on behalf of the gasless account,
     // it makes this same Paymaster a great recipient of a transaction if its only action is a pure token transfer
-    function transferToken(address target, uint256 value) external {
+    function transferToken(IERC20Metadata token, address target, uint256 value) external {
         require(msg.sender == getTrustedForwarder(), "must be a meta-tx");
-        require(token.transferFrom(_msgSender(), target, value), "transferToken failed");
+        IERC20(token).safeTransferFrom(_msgSender(), target, value);
```

Comment:
afaik, this casting is not needed: IERC20Metadata inherits from IERC20

Reply:
Fixed